hilti/toolchain/src/compiler/type-unifier.cc

```diff
@@ -203,9 +201,18 @@ class VisitorTypeUnifier : public visitor::MutatingPostOrder {
 } // namespace
 
 void type_unifier::Unifier::add(UnqualifiedType* t) {
+    // Occurs check: We cannot handle recursive types. Error out if we see the same
+    // node twice.
+    if ( _cd.haveSeen(t) ) {
+        t->addError("Invalid cycle detected in type");
```

Comment:
Suggest to drop "invalid" since we do not give a suggestion on how to do "valid" cycles. Logging the type might also be helpful, e.g.,

```suggestion
        t->addError(util::fmt("cycle detected in definition of type '%s'", t->typeID()));
```